airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java

```diff
@@ -437,14 +456,16 @@ private Duration getTimeToWait(final UUID connectionId) {
 
   /**
    * Creates a new job if it is not present in the input. If the jobId is specified in the input of
-   * the connectionManagerWorkflow, we will return it. Otherwise we will create a job and return its
-   * id.
+   * the connectionManagerWorkflow, we will return it. Otherwise we will ensure there are no other
+   * non-terminal jobs, and create a job and return its id.
    */
   private Long getOrCreateJobId(final ConnectionUpdaterInput connectionUpdaterInput) {
     if (connectionUpdaterInput.getJobId() != null) {
       return connectionUpdaterInput.getJobId();
     }
 
+    ensureCleanJobState(connectionUpdaterInput);
```

Comment:
I decided to put this "clean job state" activity call here, right before we attempt to create a new job, rather than calling it at the beginning of the connection manager workflow only when it is on its first run, as described in the issue. 

My reasons for this are as follows:
- Regardless of how we get into this state, if the connection manager workflow reaches this point in the code and there is an existing non-terminal job for this connection, the creation of a job will fail and the workflow will end up in a permanently broken state, where it tries over and over again to create a new job. A workflow being terminated and restarted while another job was already running is one way we can end up in this state (as the issue notes), but there are likely other ways that we have not accounted for.
- If we put this call at the beginning of the workflow instead, with a flag noting if it is the first run, and the connection manager workflow runs into an exception before or while it is able to execute this job state cleanup, then the workflow will be continuedAsNew from [here](https://github.com/airbytehq/airbyte/blob/07359ffd778a73e8a361d83b8b87f28b7e8fb0ab/airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java#L123), which would wipe out that `firstRun` flag and potentially put the workflow into a broken state again.

For the above reasons, it felt better to put this job cleanup logic just before we attempt to create a job, to ensure that we never end up in the permanently-broken state described above.

Reply:
Might be worth putting these reasons as a comment in the code or at least a link to this review comment so future us will remember why we chose to do it this way.  This seems like a fairly dense set of reasoning that will get lost to the sands of time if we don't document it.

Reply:
Good point, I added a comment above this line to try to capture the gist of what I said here.